private/bufpkg/bufcheck/buflint/internal/buflintcheck/buflintcheck.go

```diff
@@ -963,3 +968,96 @@ func checkSyntaxSpecified(add addFunc, file protosource.File) error {
 	}
 	return nil
 }
+
+// CheckValidateRulesTypesMatch is a check function.
+var CheckValidateRulesTypesMatch = newMessageCheckFunc(checkValidateRulesTypesMatch)
+
+var (
+	// fieldConstraintsDesc provides a Descriptor for validate.FieldConstraints.
+	fieldConstraintsDesc = (*validate.FieldConstraints)(nil).ProtoReflect().Descriptor()
+
+	// fieldConstraintsOneOfDesc provides the OneOfDescriptor for the type union
+	// in FieldConstraints.
+	fieldConstraintsOneOfDesc = fieldConstraintsDesc.Oneofs().ByName("type")
+)
+
+func checkValidateRulesTypesMatch(add addFunc, message protosource.Message) error {
+	for _, field := range message.Fields() {
+		data, err := getDataByExtension(field, validate.E_Field)
+		if err != nil {
+			return err
+		}
+		var constraints validate.FieldConstraints
+		if err := proto.Unmarshal(data, &constraints); err != nil {
+			return fmt.Errorf("could not unmarshal expected validate.disabled into validate.MessageConstraints: %v", err)
+		}
+		constraintsRefl := constraints.ProtoReflect()
+		setOneOf := constraintsRefl.WhichOneof(fieldConstraintsOneOfDesc)
+		if setOneOf == nil {
+			continue
+		}
+
+		got := field.Type()
+		expected := getType(setOneOf.Name())
+		if got != expected {
+			add(field, field.Location(), nil, "constraint type mismatch on %q, expected %q but got %s", field.FullName(), expected, got)
+		}
+	}
+
+	return nil
+}
+
+func getDataByExtension(field protosource.OptionExtensionDescriptor, extensionType protoreflect.ExtensionType) ([]byte, error) {
+	extension, ok := field.OptionExtension(extensionType)
+	if !ok {
+		return nil, nil
+	}
+	fieldConstraints, ok := extension.(*dynamicpb.Message)
+	if !ok {
+		return nil, fmt.Errorf("could not convert expected validate.disabled into bool, was %T", extension)
+	}
+	data, err := proto.Marshal(fieldConstraints)
+	if err != nil {
+		return nil, fmt.Errorf("could not marshal expected validate.disabled into bytes: %v", err)
+	}
+	return data, nil
+}
+
+func getType(in protoreflect.Name) descriptorpb.FieldDescriptorProto_Type {
```

Comment:
I'm very interested in advice on this, this feels too hand crafted but i dont know another solution